### PR TITLE
Use brave-core `AdBlockDefaultResourceProvider` in iOS

### PR DESCRIPTION
**Pull request by cuba**
Resolves https://github.com/brave/brave-browser/issues/37472

Move more ad-block code so it's based on desktop code

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

1. Test that youtube ad-blocking works during relaunch and heading directly to youtube (including additional lists such as cookie consent notices)
2. Test https://dev-pages.brave.software/filtering/cosmetic-filtering.html
3. Test https://dev-pages.brave.software/filtering/scriptlets.html
4. Test https://dev-pages.brave.software/filtering/additional-lists.html
5. Test https://dev-pages.bravesoftware.com/filtering/text-ads.html
6. Test https://dev-pages.brave.software/filtering/scriptlets.html

### CODE REVIEW

File: ios/browser/api/brave_shields/adblock_service.mm [hunk 1]
```diff
@@ -90,6 +116,7 @@ - (instancetype)initWithComponentUpdater:
         GetApplicationContext()->GetLocalState(),
         GetApplicationContext()->GetApplicationLocale(), _cus,
         _catalogProvider);
+    _resourceProvider = new brave_shields::AdBlockDefaultResourceProvider(_cus);
```
Comment (Brandon-T) [inline ios/browser/api/brave_shields/adblock_service.mm:119]: Leaking memory here.

File: ios/browser/api/brave_shields/adblock_service.mm [hunk 2]
```diff
@@ -112,13 +139,13 @@ - (void)registerCatalogChanges:(void (^)())callback {
...
 }
 
-- (void)registerResourceComponent:
-    (void (^)(NSString* _Nullable installPath))componentReady {
-  brave_shields::RegisterAdBlockDefaultResourceComponent(
-      _cus, base::BindRepeating(^(const base::FilePath& install_path) {
-        const auto installPath = base::SysUTF8ToNSString(install_path.value());
-        componentReady(installPath);
+- (void)registerResourcesChanges:(void (^)(NSString* resourcesJSON))callback {
+  _resourceObserver = new brave_shields::AdBlockResourceObserver(
```
Comment (Brandon-T) [inline ios/browser/api/brave_shields/adblock_service.mm:143]: Leaking memory here.

**Review by Brandon-T — CHANGES_REQUESTED** [review]

**Comment by Brandon-T** [comment]
This entire file is leaking memory: `ios/browser/api/brave_shields/adblock_service.mm`
You cannot call `new` without calling `delete` somewhere to clean up the memory.

Also `raw_ptr` means you're responsible for memory deallocation. If you don't want to be responsible for it, you need an ownership container like `std::unique_ptr` instead.

**Comment by cuba** [comment]
Thanks for the tips @Brandon-T , ill address the issues on both PRs.

**Review by Brandon-T — APPROVED** [review]

**Pull request merged**